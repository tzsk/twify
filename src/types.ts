import { drivers } from './drivers';

export type PackageManager = 'npm' | 'yarn' | 'pnpm';

export interface InitOptions {
  keep?: boolean;
  pretty?: boolean;
  installer?: PackageManager;
}

export interface Step {
  (): Promise<void>;
}

type CSSLocation = () => Promise<string>;

export interface Framework {
  requiredDependencies: string[];
  initCommands: string[];
  cssLocation: string | CSSLocation;
  content: string[];
  steps: Step[];
}

export type DriverImport = Promise<{ default: Framework }>;

export type Driver = keyof typeof drivers;
