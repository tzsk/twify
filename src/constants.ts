export const CSS_STUB = `@tailwind base;
@tailwind components;
@tailwind utilities;
`;

export const prettierDependencies = ['prettier', 'prettier-plugin-tailwindcss'];
