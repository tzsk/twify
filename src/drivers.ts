import { DriverImport } from './types';

export const drivers = {
  NextJS: (): DriverImport => import('./frameworks/nextjs'),
  Nuxt2: (): DriverImport => import('./frameworks/nuxt2'),
  Nuxt3: (): DriverImport => import('./frameworks/nuxt3'),
  Remix: (): DriverImport => import('./frameworks/remix'),
  SvelteKit: (): DriverImport => import('./frameworks/sveltekit'),
  Vite: (): DriverImport => import('./frameworks/vite'),
  Angular: (): DriverImport => import('./frameworks/angular'),
  CreateReactApp: (): DriverImport => import('./frameworks/cra'),
  LaravelVite: (): DriverImport => import('./frameworks/laravel'),
  Solid: (): DriverImport => import('./frameworks/solid'),
};
