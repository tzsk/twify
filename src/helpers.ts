import fs from 'fs-extra';
import { exec, spawn, SpawnOptions } from 'child_process';
import path from 'path';
import { Driver, PackageManager } from './types';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import gradient from 'gradient-string';
import chalk from 'chalk';

export function intro(warn = false) {
  console.log(gradient.fruit('\n\n🔥 Welcome to Twify!\n'));
  console.log(
    chalk.blue.bold('- A tool to help you setup your project with TailwindCSS.')
  );

  if (warn) {
    console.log(
      chalk.underline.yellow.bold(
        `\n❯ It might reconfigure any existing setup you might have.\n❯ It is advised to be used in a new project.`
      )
    );
  }
}

export function getVersion() {
  const __dirname = dirname(fileURLToPath(import.meta.url));
  const pkg = fs.readJsonSync(join(__dirname, '../package.json'));

  return pkg.version;
}

export function detectFramework(): Driver | undefined {
  const { dependencies = {}, devDependencies = {} } = fs.readJsonSync(
    path.join(process.cwd(), 'package.json')
  );

  if (dependencies['next']) return 'NextJS';
  if (dependencies['@remix-run/react']) return 'Remix';
  if (dependencies['nuxt']) return 'Nuxt2';
  if (dependencies['@angular/core']) return 'Angular';
  if (dependencies['react-scripts']) return 'CreateReactApp';
  if (devDependencies['laravel-vite-plugin']) return 'LaravelVite';
  if (devDependencies['@sveltejs/kit']) return 'SvelteKit';
  if (devDependencies['nuxt']) return 'Nuxt3';
  if (devDependencies['vite']) return 'Vite';
}

export function installerPrefix(manager?: PackageManager): string {
  const npm = 'npm install --save-dev';
  const yarn = 'yarn add --dev';
  const pnpm = 'pnpm install --save-dev';

  switch (manager) {
    case 'npm':
      return npm;
    case 'yarn':
      return yarn;
    case 'pnpm':
      return pnpm;
    default:
      return fs.existsSync(path.join(process.cwd(), 'yarn.lock')) ? yarn : npm;
  }
}

export function runCommandSpawn(command: string, options: SpawnOptions = {}) {
  return new Promise((resolve, reject) => {
    const result = spawn(command, options);

    result.on('close', async () => {
      resolve(true);
    });

    result.on('error', (err) => {
      reject(err);
    });
  });
}

export function runCommand(cmd: string) {
  return new Promise((resolve, reject) => {
    exec(cmd, { cwd: process.cwd() }, (err, stdout) => {
      if (err) {
        reject(err);
      } else {
        resolve(stdout);
      }
    });
  });
}
