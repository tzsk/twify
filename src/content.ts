import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import j from 'jscodeshift';
import { Framework } from './types';
import { tailwindConfigFiles } from './constants';

export function addContentToCode(code: string, content: string[]): string {
  const parse = j.withParser('flow');
  const ast = parse(code);

  ast
    .find(j.Property, { key: { name: 'content' } })
    .replaceWith(
      j.property(
        'init',
        j.identifier('content'),
        j.arrayExpression(content.map((c) => j.literal(c)))
      )
    );

  return ast.toSource();
}

export async function setupContent({ content }: Framework) {
  console.log(
    `\n${chalk.green('âœ”')} Configuring ${chalk.blue.bold(content)} content...`
  );
  console.log(chalk.blue(`- ${content.join('\n- ')}`));

  const [contentPath] = tailwindConfigFiles
    .map((file) => path.join(process.cwd(), file))
    .filter((file) => fs.existsSync(file));

  if (!contentPath) {
    console.log(`\n Could not find ${chalk.blue.bold('Tailwind Config')} file`);
    return;
  }

  const fileContent = await fs.readFile(contentPath, 'utf8');
  const newContent = addContentToCode(fileContent, content);

  await fs.writeFile(contentPath, newContent);
}
