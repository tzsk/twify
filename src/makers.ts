import { runCommand } from './helpers';
import { PackageManager } from './types';
import { drivers } from './drivers';

export type Maker = { cmd: string; project: keyof typeof drivers };
export type SpecialMaker = (packageManager: PackageManager) => Promise<Maker>;

const globalInstallPrefix: Record<PackageManager, string> = {
  npm: 'npm install -g',
  yarn: 'yarn global add',
  pnpm: 'pnpm install -g',
};

const specialMakes: Record<string, SpecialMaker> = {
  angular: async (manager: PackageManager) => {
    await runCommand(`${globalInstallPrefix[manager]} @angular/cli`);

    return { cmd: 'ng new', project: 'Angular' };
  },
  nuxt3: async () => {
    return { cmd: 'npx nuxi@latest init', project: 'Nuxt3' };
  },
  'nuxt-3': async () => {
    return { cmd: 'npx nuxi@latest init', project: 'Nuxt3' };
  },
  laravel: async () => {
    return {
      cmd: 'composer create-project laravel/laravel',
      project: 'LaravelVite',
    };
  },
};

const regularMakes: Record<string, Maker> = {
  vite: { cmd: 'vite', project: 'Vite' },
  svelte: { cmd: 'svelte', project: 'SvelteKit' },
  remix: { cmd: 'remix', project: 'Remix' },
  next: { cmd: 'next-app', project: 'NextJS' },
  react: { cmd: 'react-app', project: 'CreateReactApp' },
  nuxt: { cmd: 'nuxt-app', project: 'Nuxt2' },
};

export async function resolveMakeCommand(args: string[]): Promise<Maker> {
  const [manager, create, project, folder, ...options] = args;
  const extra = options.join(' ');

  const special = Object.keys(specialMakes).find((key) =>
    project.toLowerCase().includes(key)
  );

  if (special) {
    const final = await specialMakes[special](manager as PackageManager);

    return {
      cmd: `${final.cmd} ${folder} ${extra}`.trim(),
      project: final.project,
    };
  }

  const regular = Object.keys(regularMakes).find((key) =>
    project.toLowerCase().includes(key)
  );

  if (regular) {
    const app = regularMakes[regular];
    return {
      cmd: `${manager} ${create} ${app.cmd}@latest ${folder} ${extra}`.trim(),
      project: app.project,
    };
  }

  throw new Error(`Could not find a make command for ${project}`);
}

export function shouldHandleCreateCommand() {
  const postCommand = process.argv.at(2);
  if (postCommand && ['create', 'npm', 'yarn', 'pnpm'].includes(postCommand)) {
    const args = [...process.argv.slice(2)];
    if (postCommand === 'create') {
      args.unshift('npm');
    }

    return args;
  }

  return false;
}
