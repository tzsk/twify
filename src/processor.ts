import fs from 'fs-extra';
import chalk from 'chalk';
import { Framework, InitOptions } from './types';
import { CSS_STUB, prettierDependencies } from './constants';
import { installerPrefix, runCommand } from './helpers';
import { setupContent } from './content';
import ora from 'ora';

export async function handle(framework: Framework, options: InitOptions) {
  const { requiredDependencies, initCommands, cssLocation, steps } = framework;
  const installer = installerPrefix(options.installer);

  // Install required dependencies
  if (requiredDependencies.length) {
    console.log('');

    const spinner = ora('Installing Required Dependencies...').start();
    await runCommand(`${installer} ${requiredDependencies.join(' ')}`);
    spinner.succeed();

    console.log(chalk.blue(`- ${requiredDependencies.join('\n- ')}`));
  }

  // Installing Prettier Dependency
  if (options.pretty) {
    console.log('');
    const plugin = chalk.blue.bold('Tailwind Prettier Plugin');
    const spinner = ora(`Configuring ${plugin}...`).start();
    await runCommand(`${installer} ${prettierDependencies.join(' ')}`);
    spinner.succeed();

    console.log(chalk.blue(`- ${prettierDependencies.join('\n- ')}`));
  }

  // Run init commands
  if (initCommands.length) {
    console.log('');
    const spinner = ora(
      'Initializing Tailwind & PostCSS Config Files...'
    ).start();
    for (const command of initCommands) {
      await runCommand(command);
    }
    spinner.succeed();
  }

  // Write content to cssLocation
  const cssTarget =
    typeof cssLocation === 'string' ? cssLocation : await cssLocation();
  console.log(
    `\n${chalk.green('✔')} Setting up ${chalk.blue.bold(cssTarget)}...`
  );
  fs.ensureFileSync(cssTarget);
  const exitingCss = fs.readFileSync(cssTarget, 'utf8');
  const updatedCss = !options.keep ? CSS_STUB : `${exitingCss}\n\n${CSS_STUB}`;
  fs.writeFileSync(cssTarget, updatedCss);

  await setupContent(framework);

  if (steps.length) {
    // Run process
    console.log(`\n${chalk.green('✔')} Running ${steps.length} step(s)...`);
    for (const stepFunc of steps) {
      await stepFunc();
    }
  }
}
