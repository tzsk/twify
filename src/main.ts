#!/usr/bin/env node

import chalk from 'chalk';
import { Command, program } from 'commander';
import { getVersion } from './helpers';
import { InitCommand } from './commands/init';
import { CreateCommand } from './commands/create';
import { shouldHandleCreateCommand } from './makers';

export async function runApp(app: Command) {
  const input = shouldHandleCreateCommand();
  if (input) {
    return CreateCommand(input);
  }

  app
    .name(chalk.green.bold('Twify'))
    .description(chalk.blue.bold('TailwindCSS Setup Tool'))
    .version(chalk.green.bold(getVersion()));

  app
    .command('init')
    .option(
      '-i, --installer <installer>',
      'Explicitly set the package manager to use'
    )
    .option('-k, --keep', 'Keep existing CSS')
    .option('-p, --pretty', 'Configure Prettier Plugin')
    .description('Initialize TailwindCSS in the current project')
    .action((options) => InitCommand(options));

  app.parse();
}

runApp(program);
