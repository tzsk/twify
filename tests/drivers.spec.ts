import { drivers } from '../src/drivers';

describe('Drivers', () => {
  it('has a list of drivers', async () => {
    const driversList = Object.keys(drivers);
    const expectedDrivers = [
      'NextJS',
      'Nuxt2',
      'Nuxt3',
      'Remix',
      'SvelteKit',
      'Vite',
      'Angular',
      'CreateReactApp',
      'LaravelVite',
      'Solid',
    ] as (keyof typeof drivers)[];
    expect(driversList).toEqual(expectedDrivers);

    const driverList = await Promise.all(
      expectedDrivers.map((driver) => drivers[driver]())
    );
    driverList.forEach(({ default: driver }) => {
      expect(driver).toMatchSnapshot();
    });
  });
});
