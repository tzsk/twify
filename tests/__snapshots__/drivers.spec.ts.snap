// Vitest Snapshot v1

exports[`Drivers > has a list of drivers 1`] = `
{
  "content": [
    "./pages/**/*.{js,ts,jsx,tsx}",
    "./components/**/*.{js,ts,jsx,tsx}",
    "./src/**/*.{js,ts,jsx,tsx}",
    "./app/**/*.{js,ts,jsx,tsx}",
  ],
  "cssLocation": [Function],
  "initCommands": [
    "npx tailwindcss init -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [],
}
`;

exports[`Drivers > has a list of drivers 2`] = `
{
  "content": [
    "./components/**/*.{js,vue,ts}",
    "./layouts/**/*.vue",
    "./pages/**/*.vue",
    "./plugins/**/*.{js,ts}",
    "./nuxt.config.{js,ts}",
  ],
  "cssLocation": "./assets/css/main.css",
  "initCommands": [
    "npx tailwindcss init",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss@latest",
    "autoprefixer@latest",
    "@nuxt/postcss8",
  ],
  "steps": [
    [Function],
  ],
}
`;

exports[`Drivers > has a list of drivers 3`] = `
{
  "content": [
    "./components/**/*.{js,vue,ts}",
    "./layouts/**/*.vue",
    "./pages/**/*.vue",
    "./app.vue",
    "./plugins/**/*.{js,ts}",
  ],
  "cssLocation": "./assets/css/main.css",
  "initCommands": [
    "npx tailwindcss init",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss@latest",
    "autoprefixer@latest",
  ],
  "steps": [
    [Function],
  ],
}
`;

exports[`Drivers > has a list of drivers 4`] = `
{
  "content": [
    "./app/**/*.{js,ts,jsx,tsx}",
  ],
  "cssLocation": "./styles/app.css",
  "initCommands": [
    "npx tailwindcss init -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [
    [Function],
    [Function],
  ],
}
`;

exports[`Drivers > has a list of drivers 5`] = `
{
  "content": [
    "./src/**/*.{html,js,svelte,ts}",
  ],
  "cssLocation": "./src/app.css",
  "initCommands": [
    "npx tailwindcss init tailwind.config.cjs -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [
    [Function],
    [Function],
  ],
}
`;

exports[`Drivers > has a list of drivers 6`] = `
{
  "content": [
    "./index.html",
    "./src/**/*.{vue,js,ts,jsx,tsx,svelte}",
  ],
  "cssLocation": "./src/style.css",
  "initCommands": [
    "npx tailwindcss init -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [
    [Function],
  ],
}
`;

exports[`Drivers > has a list of drivers 7`] = `
{
  "content": [
    "./src/**/*.{html,ts}",
  ],
  "cssLocation": "./src/styles.css",
  "initCommands": [
    "npx tailwindcss init",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [],
}
`;

exports[`Drivers > has a list of drivers 8`] = `
{
  "content": [
    "./src/**/*.{js,jsx,ts,tsx}",
  ],
  "cssLocation": "./src/index.css",
  "initCommands": [
    "npx tailwindcss init -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [],
}
`;

exports[`Drivers > has a list of drivers 9`] = `
{
  "content": [
    "./resources/**/*.blade.php",
    "./resources/**/*.{js,jsx,ts,tsx,vue,svelte}",
  ],
  "cssLocation": "./resources/css/app.css",
  "initCommands": [
    "npx tailwindcss init -p",
  ],
  "requiredDependencies": [
    "tailwindcss",
    "postcss",
    "autoprefixer",
  ],
  "steps": [
    [Function],
  ],
}
`;
